CSPLemmas/Lemmas/Lemma17.cs

```diff
@@ -0,0 +1,166 @@
+﻿using CSP;
+using System;
+using System.Collections.Generic;
+
+namespace CSPLemmas
+{
+    public static partial class CSPLemmas
+    {
+        public static List<CspInstance> Lemma17(CspInstance instance)
+        {
+            List<CspInstance> result = new();
+            HashSet<Pair> set = findBigThreeComponent(instance.Restrictions);
+            if(set != null)
+            {
+                foreach(List<Pair> witness in findWitness(set, instance.Restrictions))
+                {
+                    CspInstance instance1 = instance.Clone();
+                    int neighbours = 0;
+                    if (instance.Restrictions.Contains(new Restriction(witness[4], witness[1]))) neighbours++;
+                    if (instance.Restrictions.Contains(new Restriction(witness[4], witness[2]))) neighbours++;
+                    if (instance.Restrictions.Contains(new Restriction(witness[4], witness[3]))) neighbours++;
+
+                    if(neighbours == 1)
+                    {
+                        CspInstance instance2 = instance.Clone();
+                        instance1.AddToResult(witness[4]);
+                        instance2.RemoveColor(witness[4]);
+                        if(instance.Restrictions.Contains(new Restriction(witness[1], witness[2])))
+                        {
+                            CspInstance instance3 = instance2.Clone();
+                            CspInstance instance4 = instance2.Clone();
+                            instance2.AddToResult(witness[0]);
+                            instance3.AddToResult(witness[1]);
+                            instance4.AddToResult(witness[2]);
```

Comment:
Zakładasz że witness[1] jest sąsiadem z witness[4]. W pseudokodzie i w pracy (v2; c2) to był po prostu sąsiad (v5;c5), ale nie wiemy który. "If (z, R) has a constraint with only one pair in the witness, say (w, R),"
Musisz chyba sprawdzić która para faktycznie sąsiaduje z witness[4] i ewentualnie zmienić kolejność żeby była na drugim miejscu